Add content dir where Ansible will look for the content provided by ACD

sivel opened the pull request:
##### SUMMARY
Add content dir where Ansible will look for content provided by ACD

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/utils/collection_loader.py

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

sivel committed:
Add content dir where Ansible will look for content provided by ACD

sivel committed:
Check all of sys.path

abadger commented:
Okay, @nitzmahone @mattclay and I talked about this today.  We discussed four possibilities with certain drawbacks:

* A special directory findable from the ansible-base installation (like the original of this PR)
  * Problem with this was that the user might install ansible-base from system packages into `/usr/lib/[..]site-packages` and collections via `pip --user` to `~/.local/lib/[...]/site-packages`. (and then distributions like Debian might throw `dist-packages` into the mix too....)

* Scan sys.path for toplevel ansible_collection directories
  * This is what we decided to do in the end since it seemed to satisfy the requirements and have the least drawbacks.
  * Drawbacks:
    * A badly written pip installed package could bash the ansible_collection namespace for third party tools. (ansible itself will be okay as it makes ansible_collections a namespace package early on.
    * A badly written pip package could overwrite other collections.  For instance, `pip install ibm` might install `ansible_collections/ibm/aix/aix_hostname.py`, and `pip install ibm_community` might contain the same filename.  (Note: none of these options prevent this; we're just trying to make accidental overwriting unlikely.)

* Scan sys.path for any toplevel which contains an ansible_collection directory (We called this, using surrogates)
  * The benefit of this is that it isolates the files from one pip installed package from another, making it less likely that files will be overwritten by mistakes in packaging.
  * Drawbacks:
    * a lot more directories would have to be scanned so there would be a slower startup time.
    * bad pip installs could still bash the namespace for third party tools
    * it would be harder for people to make the mistake of overwriting another collection's file as we would encourage separate toplevel directories but ansible would still only load the one that it encountered first in the search order.  For instance, `pip install ibm_official` installs `ibm_official/ansible_collections/ibm/aix/aix_hostname.py`.  `pip install ibm_community` installs `ibm_community/ansible_collections/ibm/aix/aix_hostname.py`.  Both files exist on the filesystem but ansible will load the ibm_community version due to sort order and ignore the ibm_official version.
    * Plugins for third party tools used by ansible-test will probably need to set PYTHONPATH to pick up the collections' code.  With this surrogate directory form, we'll end up with nested directories (`PYTHONPATH=/usr/lib/python3.7/site-packages:/usr/lib/python3.7/site-packages/ibm_official/`) which can cause problems for python.  We don't know that we'll hit any of those but it is non-standard enough that we are wary of it.

* Scan sys.path for a hardcoded toplevel surrogate directory and then adding ansible_collection directories from inside of those.  (What this PR presently does)
  * This is quicker that having arbitrary surrogate directories.
  * Drawbacks:
    * bad pip installs could still bash the namespace for third party tools
    * Since everything lives in the same toplevel namespace (in this example, `_ansible_community_distribution`, we're back to mistakes in packaging being able to overwrite files from other collections again)
    * The nested PYTHONPATH problem exists here too.

I'll update the code in this PR to reflect using `ansible_collections` as the toplevel.  nitzmahone and I both have some other coding to do with deadlines of the migration date at the beginning of March.  Whoever gets free time first will add tests and then we will be able to merge this.

abadger committed:
System installed collection paths living in the toplevel

After talking with nitzmahone and mattclay we settled on this feature
looking up collections in specifically named subdirectories of
sys.path's directories.

We discussed pros and cons of doing this in several different ways and
decided that looking for an ansible_collections package directly in
sys.path had less drawbacks than looking for ansible_collections as a
subdirectory of a package in sys.path.  In some ways of doing it, the
extra redirection could have helped point out packaging mistakes but not
enough to justify the complexity of coding it that way.

jctanner commented:
I'm a bit confused by this PR. @abadger's comment refers to some examples of installing an ibm provided collection, but the code currently -only- looks for an `ansible_collections` package ...

```
jtanner@yoga720:/mnt/c/Users/jtanner$ python -c 'import sys; print(sys.path)'
['', '/usr/lib/python2.7', '/usr/lib/python2.7/plat-x86_64-linux-gnu', '/usr/lib/python2.7/lib-tk', '/usr/lib/python2.7/lib-old', '/usr/lib/python2.7/lib-dynload', '/usr/local/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages']
jtanner@yoga720:/mnt/c/Users/jtanner$ python -c "import sys; print([os.path.join(x, 'ansible_collections') for x in sys.path])"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name 'os' is not defined
jtanner@yoga720:/mnt/c/Users/jtanner$ python -c "import sys; import os; print([os.path.join(x, 'ansible_collections') for x in sys.path])"
['ansible_collections', '/usr/lib/python2.7/ansible_collections', '/usr/lib/python2.7/plat-x86_64-linux-gnu/ansible_collections', '/usr/lib/python2.7/lib-tk/ansible_collections', '/usr/lib/python2.7/lib-old/ansible_collections', '/usr/lib/python2.7/lib-dynload/ansible_collections', '/usr/local/lib/python2.7/dist-packages/ansible_collections', '/usr/lib/python2.7/dist-packages/ansible_collections']
```

Is there a disconnect between what is being asked for and what this PR does or am I reading it?

webknjaz commented on lib/ansible/utils/collection_loader.py:
```diff
@@ -47,6 +49,12 @@ def __init__(self, config=None):
+        for path in sys.path:
+            sys_installed_path = os.path.join(os.path.abspath(path), 'ansible_collections')
+            if os.path.isdir(sys_installed_path):
+                self._n_configured_paths.append(sys_installed_path)
```
This adds paths at the end of the list. Which means that it'll be shadowed by user-provided paths. Is this intended? Should this also be documented somewhere?

nitzmahone replied:
It is definitely the intent; this feature primarily exists to support ACD and have an "it just works" fallback for 2.9-authored content to transparently resolve to, and depending on how downstream ansible-base ends up being packaged, possibly an RPM-installed version of same. User-installed content should absolutely be able to mask it.

Docs and tests are both missing from this PR at the moment, I'd argue that both need to be addressed before it's merged. That should be achievable once migration is behind us.

sivel replied:
> Docs and tests are both missing from this PR at the moment

Scratch docs. One of the caveats to allowing this feature, is that no documentation is provided for users about it.  This is purely an internal thing that we are effectively just not communicating about its existence. It will be unsupported for any users who find it and try to use it.  It's basically "for us only".

webknjaz replied:
Personally, I'd want to have it supported b/c it's easier to ship dists this way in some cases.

sivel replied:
> Personally, I'd want to have it supported b/c it's easier to ship dists this way in some cases

That decision is mostly out of my hands. The decision came from "higher" up.  The supported way of distributing collections is galaxy/ah only.

emonty replied:
Just for the record, I agree with @webknjaz - although I understand it's out of your hands. Being able to pip install a library and have that result in a corresponding ansible collection being installed would be hugely beneficial for some things. Mostly saying that in case further conversations happen by those who are higher up to register user interest in such a feature.

webknjaz replied:
@emonty FTR this is how ACD will work @ PyPI @ Ansible 2.10. So it'd be pretty easy for users to figure out how it works. Maybe it's not something that business desires, but the community would definitely want this.

webknjaz commented on lib/ansible/utils/collection_loader.py:
```diff
@@ -47,6 +49,12 @@ def __init__(self, config=None):
+        # Append all ``ansible_collections`` dirs from sys.path to the end
+        for path in sys.path:
+            sys_installed_path = os.path.join(os.path.abspath(path), 'ansible_collections')
+            if os.path.isdir(sys_installed_path):
```
How about avoiding duplicates?
```suggestion
            if (
            	os.path.isdir(sys_installed_path) and
            	sys_installed_path not in self._n_configured_paths
            ):
```

webknjaz commented on lib/ansible/utils/collection_loader.py:
```diff
@@ -12,6 +12,8 @@
 
 from types import ModuleType
 
+import ansible
```
@sivel could you please add a comment here explaining the purpose of this import?

sivel replied:
Sorry, I tried to comment like 20 times yesterday. It needs removed. It was done in the first iteration, where we were working off of the assumption that it was relative to the ansible install.

fwiw, this is now @abadger's PR

mattclay replied:
If this import is needed, please add a comment explaining why, otherwise remove it.

webknjaz reviewed (commented):
I'm curious if Ansible CLIs should list Python dists that are installed on the system in the `ansible_collections` implicit namespace. It'd be useful for the purposes of security analysis.

webknjaz commented on lib/ansible/utils/collection_loader.py:
```diff
@@ -47,6 +49,12 @@ def __init__(self, config=None):
         elif self._n_configured_paths is None:
             self._n_configured_paths = []
 
+        # Append all ``ansible_collections`` dirs from sys.path to the end
```
This comment would be a bit more useful if it explained the effect of the change:
```suggestion
        # Append all ``ansible_collections`` dirs from sys.path to the end
        # which enables the collection loader to locate and import Python
        # modules that are distributed as Python distributed packages via
        # PyPI or similar indexes. Specifically, it enables users to distribute
        # Python content as long as dists follow PEP 420 Implicit Namespace
        # Packages conventions.
        # Refs:
        # * https://packaging.python.org/guides/packaging-namespace-packages/
        # * https://packaging.python.org/guides/creating-and-discovering-plugins/#using-namespace-packages
        # * https://setuptools.readthedocs.io/en/latest/setuptools.html#namespace-packages
        # * https://setuptools.readthedocs.io/en/latest/setuptools.html#find-namespace-packages
        #
        # SECURITY CONSIDERATIONS:
        # This provides yet another place for injecting Python code.
        # We should be clear about it when informing users about the places
        # where modules/plugins are loaded from.
        #
        # NOTE: It is possible to publish a "broken" namespace package
        # NOTE: so that when it's installed it breaks user's ability
        # NOTE: to install other Python packages in that namespace.
```

abadger replied:
I'd take out the REFS.  Security considerations should go somewhere else.  maybe there should be a single place in the documentation which lists all the locations that the collections can be loaded from.  The note is correct but within ansible, we use our special loader which establishes a namespace prior to loading anything else so this should only break with third party code which might try to load the ansible-collections package without establishing the namespace first.  So if we state that here, it should note that it won't break us, only third parties.

webknjaz commented on lib/ansible/utils/collection_loader.py:
```diff
@@ -47,6 +49,12 @@ def __init__(self, config=None):
+        for path in sys.path:
+            sys_installed_path = os.path.join(os.path.abspath(path), 'ansible_collections')
+            if os.path.isdir(sys_installed_path):
+                self._n_configured_paths.append(sys_installed_path)
```
wrong path.
```suggestion
                self._n_configured_paths.append(path)
```

webknjaz committed:
Fix adding the top-level collection dir path

webknjaz commented:
So here's a demo Python dist working with this patch: https://github.com/webknjaz/ansible-collection-python-dist-boo / https://pypi.org/project/ansible-collections.python.dist.boo/ / https://test.pypi.org/project/ansible-collections.python.dist.boo/
```
$ pip install git+https://github.com/sivel/ansible@acd-content-dir
$ pip install -i https://test.pypi.org/simple/ ansible-collections.python.dist.boo --pre
$ ansible -m python.dist.boo -a name=Bob localhost
```

abadger committed:
Avoid adding duplicate search paths

Co-Authored-By: Sviatoslav Sydorenko <webknjaz@redhat.com>

abadger commented:
> I'm a bit confused by this PR. @abadger's comment refers to some examples of installing an ibm provided collection, but the code currently -only- looks for an `ansible_collections` package ...
> 
In the example, the ibm collection would be installed inside of the `ansible_collections` python namespace package.  The collection loader will find all the collections inside of there provided that the proper upper python package is put into the collection path.  (Thanks to webknjaz for fixing which path that needs to be).  So another example from the installation-and-use side:

With standard sys.path settings in Fedora and RHEL:

* A collection module is installed into `~/.local/lib/python3.8/site-packages/ansible_collections/ibm/community/plugins/modules/foo.py`
  * `ansible -m ibm.community.foo` would use that module.
* A collection module is installed into `/usr/lib64/python3.8/site-packages/ansible_collections/community/general/plugins/modules/ping.py`
  * `ansible -m community.general.ping` would use that module.

webknjaz committed:
Add site-packages collections loading tests

Co-Authored-By: Matt Clay <matt@mystile.com>

abadger commented:
@webknjaz I think the new tests are failing here:
```
02:27 + ansible -m python.dist.boo -a name=Bob -c local localhost -vvvvvv
```

on some platforms.  Maybe the script should be more verbose in the information it's outputting since I'm not sure that's the problem... it just looks like cleanup is run right after that command.

See shippable for more output from the script.

webknjaz commented:
@abadger yes, some tests fail and I have no idea why. I asked @mattclay and he pointed out that those are py3 envs, while py2 envs all succeed. This is weird because I run py3 locally.

webknjaz commented on test/integration/targets/collections_runtime_pythonpath/aliases:
```diff
@@ -0,0 +1,4 @@
+posix
```
@mattclay are the aliases correct?

mattclay replied:
The `posix` alias is not needed since it is the default.

webknjaz replied:
Okay. I've copied it from other collection tests.

webknjaz commented on test/integration/targets/collections_runtime_pythonpath/aliases:
```diff
@@ -0,0 +1,4 @@
+posix
+shippable/posix/group4
+skip/python2.6
+skip/python2.7
```
@abadger ah, py2 doesn't fail because we skip it

mattclay replied:
Why is Python 2.7 being skipped?

webknjaz replied:
No specific reason. I was playing w/ py3 only at first and wanted to make it work before going further.

webknjaz commented:
So the problem is:
```console
01:07 localhost | FAILED! => {
01:07     "msg": "The module python.dist.boo was not found in configured module paths"
01:07 }
```

`PYTHONPATH` seems to be set up properly:
```console
01:07 + PYTHONPATH=/root/ansible/test/results/.tmp/integration/collections_runtime_pythonpath-4s4es2i9-ÅÑŚÌβŁÈ/test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo:/root/ansible/lib
```

Can it be because of `ÅÑŚÌβŁÈ`?

webknjaz commented:
> Can it be because of `ÅÑŚÌβŁÈ`?

Injecting it locally doesn't break things...

webknjaz commented:
It works with `--local` but I cannot test w/ docker because it's broken on my laptop atm.

webknjaz commented:
```console
13734 1584052956.66002: plugin lookup for python.dist.boo failed; errors: No module named 'ansible_collections.python'
```

webknjaz commented:
So `os.path.isdir()` returns `False` in the test envs for some reason. But not on my laptop locally...

webknjaz commented:
Oh wow! So `ansible_collections` subdirectory isn't getting copied into VMs. Everything around it is.

@mattclay does `ansible-test` somehow exclude dirs with this name?

mattclay commented:
@webknjaz I don't see a `ansible_collections` subdir in the PR.

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/README.rst:
```diff
@@ -0,0 +1,107 @@
+.. image:: https://img.shields.io/pypi/v/ansible-collections.python.dist.boo.svg?logo=Python&logoColor=white
```
Please omit unnecessary files from the test. The additional files distract from the actual test and make it much more difficult to see what is actually being tested.

webknjaz commented:
Recent findings: turned out that our `.gitignore` matches `dist` subfolder which has lead to not including the module files. I've added them now and will try to polish this tomorrow.

webknjaz committed:
Fix a duplicate path check

webknjaz commented:
@mattclay @abadger does this test really need Python 2?

abadger commented:
Yes, it needs to work on python2 as well.  It's probably important to test there as it makes sure we haven't broken anything in regards to using our loader to emulate new style namespaces on python2

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/.gitignore:
```diff
@@ -0,0 +1,403 @@
+# Created by https://www.gitignore.io/api/git,dotenv,linux,pydev,python,windows,pycharm+all,jupyternotebooks,vim,webstorm,emacs
```
Please remove this and other files that are unnecessary for the test to function. Unnecessary content obscures what is actually being tested.

webknjaz replied:
gitignore is needed, otherwise, the git tree will be unclean

mattclay replied:
I'm not able to reproduce this. How are you running the tests and what stray files are you seeing?

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,56 @@
+    echo Cleanup: removing build artifacts...
+    rm -rvf ansible-collection-python-dist-boo/{dist,build}/
+}
+trap _cleanup INT TERM EXIT QUIT
```
This cleanup should be unnecessary, since `ansible-test` removes the temporary directory used for tests.

webknjaz replied:
I prefer to clean after myself regardless

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,56 @@
+>&2 echo \
+    Test 1: build a binary Python dist \
+    '(a wheel)' using PEP517.
+pip install pep517
```
```suggestion
pip install pep517 --disable-pip-version-check
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,56 @@
+    Test 1: install a pre-built dist \
+    with pip.
+pip install \
+  --no-index \
```
```suggestion
  --no-index --disable-pip-version-check \
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,56 @@
+  -f ansible-collection-python-dist-boo/dist/ \
+  --only-binary=ansible-collections.python.dist \
+  ansible-collections.python.dist
+pip show ansible-collections.python.dist
```
```suggestion
pip show ansible-collections.python.dist --disable-pip-version-check
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-foo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,113 @@
+# -*- coding: utf-8 -*-
```
Please use minimal modules that provide the necessary test functionality. No metadata, documentation or classes should be needed. Just a simple main function to return the required message. As with including unnecessary files, including unnecessary boilerplate only distracts from what is actually important in the test.

mattclay commented on lib/ansible/utils/collection_loader.py:
```diff
@@ -47,6 +49,15 @@ def __init__(self, config=None):
+            if (
+                    os.path.isdir(sys_installed_path)
+                    and path not in self._n_configured_paths
+            ):
```
This can be simplified and optimized:

1. Check `self._n_configured_paths` before calling `os.path.isdir` to avoid unnecessary filesystem checks.
2. Simplify the conditional to remove the intermediate variable since it is used only once.
3. Avoid calling `os.path.abspath` since it `os.path.isdir` will handle relative paths for us.

```suggestion
            if path not in self._n_configured_paths and os.path.isdir(os.path.join(path, 'ansible_collections')):
```

webknjaz committed:
Drop unused ansible import

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,62 @@
+
+
+export PIP_DISABLE_PIP_VERSION_CHECK=1
+TMP_DIR="$(mktemp -d)"
```
Use `OUTPUT_DIR` instead of creating a temporary directory.

webknjaz committed:
Optimize collection path checks

Co-Authored-By: Matt Clay <matt@mystile.com>

webknjaz commented:
This is what I'm facing currently:
```json
+ ansible -m python.dist.boo -a name=Frodo -c local localhost
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ImportError: No module named python.dist.plugins.modules
localhost | FAILED! => {
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1584375871.84-201201544261665/AnsiballZ_boo.py\", line 116, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1584375871.84-201201544261665/AnsiballZ_boo.py\", line 108, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1584375871.84-201201544261665/AnsiballZ_boo.py\", line 54, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.python.dist.plugins.modules.boo', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python2.7/runpy.py\", line 170, in run_module\n    mod_name, loader, code, fname = _get_module_details(mod_name)\n  File \"/usr/lib64/python2.7/runpy.py\", line 101, in _get_module_details\n    loader = get_loader(mod_name)\n  File \"/usr/lib64/python2.7/pkgutil.py\", line 464, in get_loader\n    return find_loader(fullname)\n  File \"/usr/lib64/python2.7/pkgutil.py\", line 474, in find_loader\n    for importer in iter_importers(fullname):\n  File \"/usr/lib64/python2.7/pkgutil.py\", line 430, in iter_importers\n    __import__(pkg)\nImportError: No module named python.dist.plugins.modules\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", 
    "rc": 1
}
```

I've extracted that traceback in a human-readable form:
```python-traceback
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1584375871.84-201201544261665/AnsiballZ_boo.py", line 116, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1584375871.84-201201544261665/AnsiballZ_boo.py", line 108, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1584375871.84-201201544261665/AnsiballZ_boo.py", line 54, in invoke_module
    runpy.run_module(mod_name='ansible_collections.python.dist.plugins.modules.boo', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python2.7/runpy.py", line 170, in run_module
    mod_name, loader, code, fname = _get_module_details(mod_name)
  File "/usr/lib64/python2.7/runpy.py", line 101, in _get_module_details
    loader = get_loader(mod_name)
  File "/usr/lib64/python2.7/pkgutil.py", line 464, in get_loader
    return find_loader(fullname)
  File "/usr/lib64/python2.7/pkgutil.py", line 474, in find_loader
    for importer in iter_importers(fullname):
  File "/usr/lib64/python2.7/pkgutil.py", line 430, in iter_importers
    __import__(pkg)
ImportError: No module named python.dist.plugins.modules
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,62 @@
+
+
+export PIP_DISABLE_PIP_VERSION_CHECK=1
+TMP_DIR=`mktemp -d`
```
Use the existing `OUTPUT_DIR` managed by `ansible-test` instead of creating a temp dir in the test.

webknjaz replied:
I need it outside of the path with a Unicode name because python 2 setuptools explode otherwise.

mattclay replied:
What issue are you seeing? I didn't have any problems with the test using `OUTPUT_DIR`.

webknjaz replied:
Ah, it was failing for in-place builds. Seems to be okay with what you suggest.

webknjaz commented:
There's this warning @ T=linux/centos6/4
```console
00:46 /root/ansible/lib/ansible/utils/collection_loader.py:52: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
00:46   if path not in self._n_configured_paths and os.path.isdir(os.path.join(path, 'ansible_collections')):
```
(https://app.shippable.com/github/ansible/ansible/runs/162488/106/console)

abadger committed:
Properly normalize _n_configured_paths

webknjaz committed:
Only expand make native config-provided paths

webknjaz committed:
Pass bytes to isdir to be safe under POSIX

webknjaz committed:
Add a change fragment

webknjaz commented:
I think it's ready. CI failures look unrelated.

jamescassell commented:
@ansibullbot aren't you supposed to be a bot?

abadger commented:
yes... it is... I had to generate a token for the account and didn't log out before posting that comment....

abadger reviewed (approved):
I can't figure out how to delete the review from ansibullbot which I mistakenly added.  Approving from my real account.  Code looks right and the tests are testing the use case we want.

mattclay commented on lib/ansible/utils/collection_loader.py:
```diff
@@ -38,17 +38,33 @@
+        paths = [
+            to_native(os.path.expanduser(p), errors='surrogate_or_strict')
+            for p in paths
+        ]
```
This line doesn't need to be wrapped. It was under the column limit before, and it's even shorter now.

webknjaz replied:
I don't always use the line length as a metric to decide if I should go multiline. In this case, I wanted to separate nested expressions for easier consumption by the reader.
The fact that the linter has a huge line length setting doesn't mean that we should be trying to hit that limit.

mattclay commented on lib/ansible/utils/collection_loader.py:
```diff
@@ -38,17 +38,33 @@
+                        os.path.join(path, 'ansible_collections'),
+                        errors='surrogate_or_strict',
+                    ))
+            ):
```
While this line does need to be wrapped, 2 lines would suffice instead of 7.

webknjaz replied:
There's a lot of nested calls here that make the cognitive complexity higher. It's more readable when separated into logical parts. Plus it fits better on the screen.

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
```
Remove the unnecessary boilerplate from this module. It doesn't need a copyright notice, nor does it need any classes. A simple main function, much like the ping module, is all that is needed. Doing that will cut this module down to about 25% of its current size and make it much more obvious what is going on in the test.

The same applies to the other test module.

webknjaz replied:
I wanted to avoid adding sanity ignore lines

mattclay replied:
Sanity ignore entries will not be needed.

Module specific sanity tests such as validate-modules do not apply to test modules.

webknjaz replied:
Yep, I already see it. Something else failed then.

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/setup.cfg:
```diff
@@ -0,0 +1,74 @@
+[bdist_wheel]
```
Remove the unnecessary boilerplate in this file, keeping only what is needed for the test. Doing so will make it much more readable and will help avoid obscuring the important parts of the test. At a quick glance this includes removing all the metadata except name and version, as well as install_requires and python_requires. There may be more that is unnecessary, but that's the bulk of it.

Again, this helps cut down the size of the test, leaving only the important parts needed for the test. Less to read makes the test much easier to comprehend.

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env bash
+
+set -mvEeuxo pipefail
```
Remove the extra options to cut down on the verbosity of the test. This will make the test output much easier to follow by removing unnecessarily verbose output, which is nearly half the output of the test.

```suggestion
set -eux -o pipefail
```

webknjaz replied:
That's just `-v`, why do you want to cut off others?

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,58 @@
+    echo Cleanup: uninstalling ansible-collections.python.dist...
+    python -m pip.__main__ uninstall -y ansible-collections.python.dist
+}
+trap _cleanup INT TERM EXIT QUIT
```
Remove this cleanup so that the contents of `OUTPUT_DIR` can be inspected if needed after the test ends. The cleanup is redundant, since ansible-test ensures the directory is clean before running each test. The contents are intentionally left behind after each test for easier troubleshooting.

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,58 @@
+trap _cleanup INT TERM EXIT QUIT
+
+
+>&2 echo Init: creating a virtualenv...
```
No need to echo a message here, the output from `-x` already tells us what's going on.

mattclay commented on test/integration/targets/collections_runtime_pythonpath/runme.sh:
```diff
@@ -0,0 +1,58 @@
+
+
+>&2 echo \
+    Test 0: test that the module \
```
The test numbers should be unique -- instead of 1 through 5 they're currently 0, 1, 1, 1, 2.

webknjaz replied:
Because there are 3 subtests, not 5. I grouped them: when multiple log lines correspond to the same test, I use the same number.

mattclay replied:
Unfortunately the grouping isn't obvious. When running the tests locally I ended up with a failure, saw the number on the failure, and quickly went to the part of the test I thought was responsible. It took me a while to realize I was looking at the wrong part of the test because the numbers were not unique.

If they're going to be grouped, I recommend echoing a group heading above each group rather than numbering the tests.

webknjaz replied:
I'm not sure what exactly you want me to do. Feel free to send suggested changes.

mattclay replied:
GitHub suggestions doesn't allow adding lines where no changes already exist, so I'll try explaining here instead:

1. Remove the numbers from each test label, or make them unique.
2. If grouping of the tests is desired, echo a line with the group name/description before the start of each group.

mattclay reviewed (dismissed):
Functionally the changes look OK. However, the PR should be trimmed down so it's easier to see what's going on in the test. Instead of the 1400+ lines currently in the PR it should probably be under 200. There are several unnecessary files that should be removed:

- .gitignore - There are no files created in the source tree by the test.
- LICENSE - Not used by the test.
- README.rst - Not used by the test.

webknjaz commented:
> LICENSE - Not used by the test.
> README.rst - Not used by the test.

These are required to have a correct dist. I don't want it to be half-baked.

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/setup.cfg:
```diff
@@ -0,0 +1,53 @@
+    Topic :: Software Development :: Libraries
+    Topic :: Software Development :: Libraries :: Python Modules
+keywords =
+    Ansible Collections
```
This metadata is not required by the test. The test should focus on what is needed to verify collection loading. This looks more like an example of how to publish a collection for PyPI, which is something that should go in an example repo or docs rather than a test.

```suggestion
```

webknjaz replied:
I don't like the idea of having a broken dist in tests.

mattclay replied:
How is it broken?

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/setup.cfg:
```diff
@@ -0,0 +1,53 @@
+    Ansible Collections
+
+[options]
+python_requires = >=2.7
```
Not required by the test.

```suggestion
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/setup.cfg:
```diff
@@ -0,0 +1,53 @@
+
+# These are required in actual runtime:
+install_requires =
+    ansible
```
Not required by the test.

```suggestion
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-foo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,32 @@
+    )
+    name = module.params['name']
+    if not name:
+        module.fail_json(msg='`name` arg should not be empty')
```
Not required by the test.

```suggestion
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-foo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,32 @@
+            'name': {'default': 'світ'},
+        },
+        required_one_of=[['name']],
+        supports_check_mode=False,
```
Not required by the test.

```suggestion
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,32 @@
+    )
+    name = module.params['name']
+    if not name:
+        module.fail_json(msg='`name` arg should not be empty')
```
Not required by the test.

```suggestion
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,32 @@
+            'name': {'default': 'world'},
+        },
+        required_one_of=[['name']],
+        supports_check_mode=False,
```
Not required by the test.

```suggestion
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,32 @@
+from ansible.module_utils.basic import AnsibleModule
+
+
+def execute_boo():
```
Use the standard naming convention for the main function.

```suggestion
def main():
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-boo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,32 @@
+
+
+if __name__ == '__main__':
+    execute_boo()
```
```suggestion
    main()
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-foo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,32 @@
+from ansible.module_utils.basic import AnsibleModule
+
+
+def execute_boo():
```
Use the standard naming convention for the main function.

```suggestion
def main():
```

mattclay commented on test/integration/targets/collections_runtime_pythonpath/ansible-collection-python-dist-foo/ansible_collections/python/dist/plugins/modules/boo.py:
```diff
@@ -0,0 +1,32 @@
+
+
+if __name__ == '__main__':
+    execute_boo()
```
```suggestion
    main()
```

mattclay reviewed (commented):
> > LICENSE - Not used by the test.
> > README.rst - Not used by the test.
> 
> These are required to have a correct dist. I don't want it to be half-baked.

These files (as well as `.gitignore`) can be removed from the test with no impact to its functionality.  What purpose do they serve?

mattclay reviewed (dismissed):
Last review should have requested changes, not been a comment.

mattclay commented on changelogs/fragments/acd-content-dir.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- ACD - Add content dir where Ansible will look for content provided by ACD
```
Remove this fragment. The first fragment already provides a good description of the changes.

webknjaz replied:
Oh, I missed that sivel added this earlier

webknjaz committed:
Drop unnecessary change fragment

webknjaz commented:
Ah, GH lost my commit message and co-authors on merge :(
